java/tables/src/main/java/sleeper/table/index/dynamodb/DynamoDBTableIndex.java

```diff
@@ -0,0 +1,129 @@
+
+    public static final String TABLE_NAME_FIELD = DynamoDBTableIdFormat.TABLE_NAME_FIELD;
+    public static final String TABLE_ID_FIELD = DynamoDBTableIdFormat.TABLE_ID_FIELD;
+
+    private final AmazonDynamoDB dynamoDB;
+    private final String nameIndexDynamoTableName;
+    private final String idIndexDynamoTableName;
+    private final TableIdGenerator idGenerator = new TableIdGenerator();
+
+    public DynamoDBTableIndex(AmazonDynamoDB dynamoDB, InstanceProperties instanceProperties) {
+        this.dynamoDB = dynamoDB;
+        this.nameIndexDynamoTableName = instanceProperties.get(TABLE_NAME_INDEX_DYNAMO_TABLENAME);
+        this.idIndexDynamoTableName = instanceProperties.get(TABLE_ID_INDEX_DYNAMO_TABLENAME);
+    }
+
+    @Override
+    public TableId createTable(String tableName) throws TableAlreadyExistsException {
+        TableId id = TableId.idAndName(idGenerator.generateString(), tableName);
+        TransactWriteItemsRequest request = new TransactWriteItemsRequest()
+                .withReturnConsumedCapacity(ReturnConsumedCapacity.TOTAL)
+                .withTransactItems(
+                        new TransactWriteItem().withPut(new Put()
+                                .withTableName(nameIndexDynamoTableName)
+                                .withItem(DynamoDBTableIdFormat.getItem(id))
+                                .withConditionExpression("attribute_not_exists(#tablename)")
+                                .withExpressionAttributeNames(Map.of("#tablename", TABLE_NAME_FIELD))),
+                        new TransactWriteItem().withPut(new Put()
+                                .withTableName(idIndexDynamoTableName)
+                                .withItem(DynamoDBTableIdFormat.getItem(id))
+                                .withConditionExpression("attribute_not_exists(#tableid)")
+                                .withExpressionAttributeNames(Map.of("#tableid", TABLE_ID_FIELD))));
+        try {
+            TransactWriteItemsResult result = dynamoDB.transactWriteItems(request);
+            List<ConsumedCapacity> consumedCapacity = result.getConsumedCapacity();
+            double totalCapacity = consumedCapacity.stream().mapToDouble(ConsumedCapacity::getCapacityUnits).sum();
+            LOGGER.debug("Created table {} with ID {}, capacity consumed = {}",
+                    tableName, id.getTableId(), totalCapacity);
+            return id;
+        } catch (TransactionCanceledException e) {
+            CancellationReason nameIndexReason = e.getCancellationReasons().get(0);
+            if ("ConditionalCheckFailed".equals(nameIndexReason.getCode())) {
+                throw new TableAlreadyExistsException(tableName);
+            } else {
+                throw e;
+            }
+        }
+    }
+
+    @Override
+    public Stream<TableId> streamAllTables() {
+        return streamPagedItems(dynamoDB, new ScanRequest().withTableName(nameIndexDynamoTableName))
```

Comment:
We should probably use strongly consistent reads.

Reply:
Done